File: eeg_analysis/r/eeg_basic_analysis.R
# ADS1299-Core-SDK: beginner EEG analysis in base R
# Data convention: samples x channels, amplitudes in microvolts (uV).
# This is research/education starter code, not a clinical processing standard.

ads1299_eeg_basic <- function(x, fs = 250, mains_hz = 50) {
  x <- as.matrix(x)
  if (nrow(x) < 10 || ncol(x) < 1) stop("x must be samples x channels")
  if (fs <= 0) stop("fs must be positive")

  # Demean each channel.
  x <- sweep(x, 2, colMeans(x, na.rm = TRUE), "-")

  # Common-average reference when at least two channels are present.
  if (ncol(x) >= 2) {
    car <- rowMeans(x, na.rm = TRUE)
    x <- x - car
  }

  # Simple FFT-based PSD helper. For publication work, use a validated Welch
  # implementation (e.g. signal package) and document all parameters.
  psd_one <- function(v) {
    n <- length(v)
    w <- 0.5 - 0.5 * cos(2*pi*(0:(n-1))/(n-1))
    y <- fft(v * w)
    k <- 0:floor(n/2)
    p <- (Mod(y[k + 1])^2) / (fs * sum(w^2))
    if (length(p) > 2) p[2:(length(p)-1)] <- 2*p[2:(length(p)-1)]
    list(freq = k*fs/n, psd = p)
  }

  psd <- lapply(seq_len(ncol(x)), function(ch) psd_one(x[, ch]))
  freq <- psd[[1]]$freq
  pmat <- do.call(cbind, lapply(psd, `[[`, "psd"))

  band_power <- function(lo, hi) {
    idx <- which(freq >= lo & freq < hi)
    if (length(idx) < 2) return(rep(NA_real_, ncol(x)))
    df <- mean(diff(freq[idx]))
    colSums(pmat[idx, , drop = FALSE]) * df
  }

  bands <- list(
    delta = c(0.5, 4), theta = c(4, 8), alpha = c(8, 13),
    beta = c(13, 30), gamma = c(30, min(45, fs/2 - 0.01))
  )
  abs_power <- lapply(bands, function(b) band_power(b[1], b[2]))
  names(abs_power) <- names(bands)

  alpha_idx <- which(freq >= 8 & freq <= 13)
  alpha_peak_hz <- sapply(seq_len(ncol(x)), function(ch) {
    if (!length(alpha_idx)) return(NA_real_)
    freq[alpha_idx[which.max(pmat[alpha_idx, ch])]]
  })

  list(
    fs = fs,
    referenced_uV = x,
    freq_Hz = freq,
    psd_uV2_per_Hz = pmat,
    bandpower_absolute = abs_power,
    alpha_peak_Hz = alpha_peak_hz,
    note = paste("mains_hz =", mains_hz, "is metadata only in this base-R starter; add a validated notch/filter before scientific use")
  )
}

File: algorithms/matlab/eeg_basic_analysis.m
function result = eeg_basic_analysis(data, fs, mainsHz)
%EEG_BASIC_ANALYSIS Beginner EEG workflow for ADS1299-Core-SDK.
%
% Data convention: samples x channels, amplitudes in microvolts (uV).
% Requires MATLAB Signal Processing Toolbox for iirnotch, butter, filtfilt,
% and pwelch.
%
% Example with synthetic data:
%   result = eeg_basic_analysis();
%
% Example with real data:
%   result = eeg_basic_analysis(eeg_uV, 250, 50);
%
% This is research/education starter code, not a clinical processing standard.

if nargin < 1 || isempty(data)
    fs = 250;
    mainsHz = 50;
    duration = 20;
    t = (0:(fs*duration-1))' / fs;
    rng(7);
    data = zeros(length(t), 8);
    for ch = 1:8
        alphaHz = 9.5 + 0.2*(ch-1);
        data(:,ch) = ...
            (15 + ch).*sin(2*pi*alphaHz*t) + ...
            5.*sin(2*pi*6*t + 0.15*(ch-1)) + ...
            8.*sin(2*pi*mainsHz*t) + ...
            6.*randn(size(t));
    end
end

if nargin < 2 || isempty(fs)
    fs = 250;
end
if nargin < 3 || isempty(mainsHz)
    mainsHz = 50;
end

validateattributes(data, {'numeric'}, {'2d','nonempty','real'});
validateattributes(fs, {'numeric'}, {'scalar','positive'});
if mainsHz >= fs/2
    error('mainsHz must be below Nyquist.');
end

% 1) Remove linear trend channel-by-channel.
x = detrend(double(data), 'linear');

% 2) Common average reference (CAR).
if size(x,2) >= 2
    x = x - mean(x, 2, 'omitnan');
end

% 3) Zero-phase notch.
wo = mainsHz/(fs/2);
bw = wo/30; % approximate Q=30
[bn, an] = iirnotch(wo, bw);
x = filtfilt(bn, an, x);

% 4) Zero-phase 0.5-45 Hz Butterworth band-pass.
lowHz = 0.5;
highHz = min(45, fs/2 - 0.5);
[bpB, bpA] = butter(4, [lowHz highHz]/(fs/2), 'bandpass');
x = filtfilt(bpB, bpA, x);

% 5) Welch PSD.
nfftWindow = min(size(x,1), max(32, round(2*fs)));
noverlap = floor(nfftWindow/2);
for ch = 1:size(x,2)
    [p(:,ch), f] = pwelch(x(:,ch), hann(nfftWindow), noverlap, [], fs); %#ok<AGROW>
end

% 6) Canonical band powers.
bands = struct( ...
    'delta', [0.5 4], ...
    'theta', [4 8], ...
    'alpha', [8 13], ...
    'beta',  [13 30], ...
    'gamma', [30 min(45, fs/2-0.5)]);

names = fieldnames(bands);
totalMask = f >= 0.5 & f <= min(45, fs/2-0.5);
totalPower = trapz(f(totalMask), p(totalMask,:), 1);
absolute = struct();
relative = struct();
for k = 1:numel(names)
    name = names{k};
    range = bands.(name);
    mask = f >= range(1) & f < range(2);
    if nnz(mask) >= 2
        pow = trapz(f(mask), p(mask,:), 1);
    else
        pow = nan(1,size(x,2));
    end
    absolute.(name) = pow;
    relative.(name) = pow ./ totalPower;
end

% 7) Alpha peak per channel.
alphaMask = f >= 8 & f <= 13;
alphaFreqs = f(alphaMask);
alphaPsd = p(alphaMask,:);
[~, idx] = max(alphaPsd, [], 1);
alphaPeakHz = alphaFreqs(idx);

% 8) Time-domain features.
timeFeatures.mean = mean(x, 1, 'omitnan');
timeFeatures.std = std(x, 0, 1, 'omitnan');
timeFeatures.rms = sqrt(mean(x.^2, 1, 'omitnan'));
timeFeatures.peakToPeak = max(x, [], 1) - min(x, [], 1);

% 9) Hjorth parameters.
dx = diff(x,1,1);
ddx = diff(dx,1,1);
var0 = var(x,0,1);
var1 = var(dx,0,1);
var2 = var(ddx,0,1);
mobility = sqrt(var1 ./ var0);
mobilityDx = sqrt(var2 ./ var1);
complexity = mobilityDx ./ mobility;

hjorth.activity = var0;
hjorth.mobility = mobility;
hjorth.complexity = complexity;

result.fs = fs;
result.filtered_uV = x;
result.freq_Hz = f;
result.psd_uV2_per_Hz = p;
result.bandpower_absolute = absolute;
result.bandpower_relative = relative;
result.alphaPeak_Hz = alphaPeakHz;
result.timeFeatures = timeFeatures;
result.hjorth = hjorth;

fprintf('ADS1299-Core-SDK MATLAB EEG starter\n');
fprintf('Samples: %d, channels: %d, fs: %.1f Hz\n', size(x,1), size(x,2), fs);
fprintf('Alpha peak (Hz): ');
fprintf('%.2f ', alphaPeakHz);
fprintf('\n');
end
